// @chantouchsek/jwt-redis/dist/JwtRedis.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const jsonwebtoken = require("jsonwebtoken");
const Redis_1 = require("./Redis");
const TokenInvalidError_1 = require("./error/TokenInvalidError");
const TokenDestroyedError_1 = require("./error/TokenDestroyedError");
const utils_1 = require("./utils");
class JwtRedis {
    constructor(client, options) {
        this.client = client;
        this.options = Object.assign({ prefix: 'jwt_label:' }, options || {});
        this.redis = new Redis_1.default(client);
    }
    sign(payload, secretOrPrivateKey, options) {
        return Promise.resolve()
            .then(async () => {
            const jti = payload.jti || utils_1.generateId(10);
            const token = jsonwebtoken.sign(Object.assign(Object.assign({}, payload), { jti }), secretOrPrivateKey, options);
            const decoded = jsonwebtoken.decode(token);
            const key = this.options.prefix + jti;
            if (decoded.exp) {
                await this.redis.setExpire(key, 'true', 'EX', Math.floor(decoded.exp - Date.now() / 1000));
            }
            else {
                await this.redis.set(key, 'true');
            }
            return token;
        });
    }
    destroy(jti) {
        const key = this.options.prefix + jti;
        return this.redis.del(key);
    }
    decode(token, options) {
        return jsonwebtoken.decode(token, options);
    }
    verify(token, secretOrPublicKey, options) {
        return new Promise((resolve, reject) => {
            return jsonwebtoken.verify(token, secretOrPublicKey, options, (err, decoded) => {
                if (err) {
                    reject(err);
                }
                resolve(decoded);
            });
        }).then((decoded) => {
            if (!decoded.jti) {
                throw new TokenInvalidError_1.default();
            }
            const { jti } = decoded;
            const key = this.options.prefix + jti;
            return this.redis.get(key)
                .then((result) => {
                if (!result) {
                    throw new TokenDestroyedError_1.default();
                }
                return decoded;
            });
        });
    }
}
exports.default = JwtRedis;
